jetty-openid/src/main/java/org/eclipse/jetty/security/openid/OpenIdCredentials.java

```diff
@@ -0,0 +1,201 @@
+import org.eclipse.jetty.util.log.Log;
+import org.eclipse.jetty.util.log.Logger;
+
+public class OpenIdCredentials
+{
+    private static final Logger LOG = Log.getLogger(OpenIdCredentials.class);
+
+    private final String redirectUri;
+    private final OpenIdConfiguration configuration;
+    private String authCode;
+    private Map<String, Object> response;
+    private Map<String, Object> claims;
+
+    public OpenIdCredentials(String authCode, String redirectUri, OpenIdConfiguration configuration)
+    {
+        this.authCode = authCode;
+        this.redirectUri = redirectUri;
+        this.configuration = configuration;
+    }
+
+    public String getUserId()
+    {
+        return (String)claims.get("sub");
+    }
+
+    public Map<String, Object> getClaims()
+    {
+        return claims;
+    }
+
+    public Map<String, Object> getResponse()
+    {
+        return response;
+    }
+
+    public void redeemAuthCode() throws IOException
+    {
+        if (LOG.isDebugEnabled())
+            LOG.debug("redeemAuthCode() {}", this);
+
+        if (authCode != null)
+        {
+            try
+            {
+                response = claimAuthCode(authCode);
+                if (LOG.isDebugEnabled())
+                    LOG.debug("response: {}", response);
+
+                String idToken = (String)response.get("id_token");
+                if (idToken == null)
+                    throw new IllegalArgumentException("no id_token");
+
+                String accessToken = (String)response.get("access_token");
+                if (accessToken == null)
+                    throw new IllegalArgumentException("no access_token");
+
+                String tokenType = (String)response.get("token_type");
+                if (!"Bearer".equalsIgnoreCase(tokenType))
+                    throw new IllegalArgumentException("invalid token_type");
+
+                claims = decodeJWT(idToken);
+                if (LOG.isDebugEnabled())
+                    LOG.debug("claims {}", claims);
+                validateClaims();
+            }
+            finally
+            {
+                // reset authCode as it can only be used once
+                authCode = null;
+            }
+        }
+    }
+
+    private void validateClaims()
+    {
+        // Issuer Identifier for the OpenID Provider MUST exactly match the value of the iss (issuer) Claim.
+        if (!configuration.getIssuer().equals(claims.get("iss")))
+        {
+            LOG.warn("Invalid \"iss\" Claim: was {}, expected {}", claims.get("iss"), configuration.getIssuer());
+            throw new IllegalArgumentException("invalid iss claim");
+        }
+
+        // The aud (audience) Claim MUST contain the client_id value.
+        if (!configuration.getClientId().equals(claims.get("aud")))
+        {
+            LOG.warn("Audience Claim MUST contain the client_id value");
+            throw new IllegalArgumentException("invalid aud claim");
```

Comment:
Don't WARN _and_ throw, it's just too noisy. 

Either DEBUG (if there is more information to log) and throw, or just throw.

Here, I would say just throw and improve the exception message (with the WARN message). 

Fix also the other places where you warn and throw.